PYIC-5685: Write to session credentials table

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Write to session credentials table

### Why did it change

[PYIC-5685: Store credentials in SessionCredentialsTable](https://github.com/govuk-one-login/ipv-core-back/commit/7e49c33d4b083e9106c01dc4a3d6ba558fd63313)

This adds logic to start storing VCs received in the new
SessionCredentialsTable. It creates a new service to do this, and wires
it into the CriStoringService.

It also updates our VerifiableCredential class to store the parsed
credential alongside the raw format. This will allow us to easily access
the signature when needed. We've already parsed it so might as well keep
it.



[PYIC-5685: Store VC in session credential table on reuse](https://github.com/govuk-one-login/ipv-core-back/commit/39471a56543b19ec468943fa9e9edfb812697030) 

If a user has already proved their identity, we want to move their VC to
the session credentials table. This is so that when we come to build
their identity, the VC required to do so are in the same place as if
they'd just completed a full identity proving journey.

[PYIC-5685: Remove creds from session credential store in reset-identity](https://github.com/govuk-one-login/ipv-core-back/commit/1b26cd3dd57197435e2894f6990454305ebfa8e5)

This updates the reset-identity handler to remove any VC from the
session credential table for the users current session, as well as from
the tactical store.

I'm a bit dubious that there will ever actually be anything to remove
here. If we're resetting the users identity they'll have failed to match
a profile and so won't have anything in the session credentials store
yet. And this happens at the start of their journey, so they shouldn't
have visited any CRI yet.

[PYIC-5685: Feature flag writes to session-credentials table](https://github.com/govuk-one-login/ipv-core-back/pull/1800/commits/aeda0ccba5e03346a67f0f0c5a5f7c3589b21104)

We need sign off from security before we can start writing to this new
table in prod. This puts the writes behind a flag which we'll be able to
control in lower envs in the usual way.

[PYIC-5685: Write VC to session credentials table for 6MFC](https://github.com/govuk-one-login/ipv-core-back/pull/1800/commits/b17c9c442ad5438e8413bd9278beb6a676551478)

When a user is sent on the 6 month fraud check journey we need to ensure
that we have all their existing VC, except for the fraud one, in the
session credentials table. This is so we can do the usual GPG45
calculations once they've collected their new fraud VC.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5685](https://govukverify.atlassian.net/browse/PYIC-5685)


[PYIC-5685]: https://govukverify.atlassian.net/browse/PYIC-5685?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-5685: Store credentials in SessionCredentialsTable

This adds logic to start storing VCs received in the new
SessionCredentialsTable. It creates a new service to do this, and wires
it into the CriStoringService.

It also updates our VerifiableCredential class to store the parsed
credential alongside the raw format. This will allow us to easily access
the signature when needed. We've already parsed it so might as well keep
it.

Wynndow committed:
PYIC-5685: Store VC in session credential table on reuse

If a user has already proved their identity, we want to move their VC to
the session credentials table. This is so that when we come to build
their identity, the VC required to do so are in the same place as if
they'd just completed a full identity proving journey.

Wynndow committed:
PYIC-5685: Remove creds from session credential store in reset-identity

This updates the reset-identity handler to remove any VC from the
session credential table for the users current session, as well as from
the tactical store.

I'm a bit dubious that there will ever actually be anything to remove
here. If we're resetting the users identity they'll have failed to match
a profile and so won't have anything in the session credentials store
yet. And this happens at the start of their journey, so they shouldn't
have visited any CRI yet.

Wynndow committed:
PYIC-5685: Feature flag writes to session-credentials table

We need sign off from security before we can start writing to this new
table in prod. This puts the writes behind a flag which we'll be able to
control in lower envs in the usual way.

Wynndow committed:
PYIC-5685: Write VC to session credentials table for 6MFC

When a user is sent on the 6 month fraud check journey we need to ensure
that we have all their existing VC, except for the fraud one, in the
session credentials table. This is so we can do the usual GPG45
calculations once they've collected their new fraud VC.

Joe-Edwards-GDS commented on lambdas/call-ticf-cri/src/main/java/uk/gov/di/ipv/core/callticfcri/CallTicfCriHandler.java:
```diff
@@ -75,6 +76,7 @@ public CallTicfCriHandler() {
                         new AuditService(AuditService.getDefaultSqsClient(), configService),
                         null,
                         new VerifiableCredentialService(configService),
+                        new SessionCredentialsService(configService),
```
Is it irrational to be annoyed that one is plural and one is singular 👀

Wynndow replied:
Not at all. But I'd be annoyed with the original one...

Joe-Edwards-GDS commented on libs/verifiable-credentials/src/main/java/uk/gov/di/ipv/core/library/verifiablecredential/service/SessionCredentialsService.java:
```diff
@@ -0,0 +1,83 @@
+    public void persistCredentials(List<VerifiableCredential> vcs, String ipvSessionId)
+            throws VerifiableCredentialException {
+        for (var vc : vcs) {
+            persistCredential(vc, ipvSessionId, false);
```
It feels a bit surprising that this sets `receivedThisSession = false` but the singular method requires you to specify.

Can we add the flag here as well, or create distinct methods for each type (e.g. `persistNewCredential` vs. `persistExistingCredential`)?

Wynndow replied:
Yep, this was mostly for convenience (when would we be storing multiple VC received this session?), but you're right it's probably better to be explicit.

Joe-Edwards-GDS commented on libs/verifiable-credentials/src/main/java/uk/gov/di/ipv/core/library/verifiablecredential/service/SessionCredentialsService.java:
```diff
@@ -0,0 +1,83 @@
+                        credential.toSessionCredentialItem(ipvSessionId, receivedThisSession),
+                        ConfigurationVariable.SESSION_CREDENTIALS_TTL);
+            } else {
+                LOGGER.info(
```
(might this be better as `debug`?)

Wynndow replied:
To be honest, if it's getting moved to `debug`, i'll probably just remove it.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -444,16 +456,22 @@ private JourneyResponse buildReuseResponse(
                     LogHelper.buildLogMessage(
                             "Fraud VC found and expired, resetting identity for GPG45 evaluation."));
+            sessionCredentialsService.persistCredentials(
+                    allVcsExceptFraud(vcs), auditEventUser.getSessionId());
```
In this scenario, I think we'll have already persisted _all_ the credentials in `checkForProfileMatch`?

Wynndow replied:
Ooo yeah. So we need to remove the fraud VC, rather than add everything except. Althogh that raises the question of if we should be allowing expired VC to contribute when calculating gpg45 scores?

Joe-Edwards-GDS replied:
It's a good question... Strictly speaking we probably should exclude it (since it's no longer a valid fraud score), but it's quite convenient to allow it...
- We want to check if you meet a GPG45 profile with the expired fraud VC before allowing you to take a repeat fraud check journey
- We are testing this by getting the fraud CRI to generate an expired fraud VC as part of a journey

Wynndow committed:
PYIC-5685: Make receivedThisSession explicit

This enforces specifying if credentials were receieved during this
session or not when storing multiple credentials. Whilst we're most
often going to be storing credentials received in a previous session in
bulk, it's good to be obvious about it.

Wynndow committed:
PYIC-5685: Refactor reuse credential persistance

This moves the logic that stores the credentials in a reuse journey
closer to where we determine which type of reuse journey the user is on.

This allows us to be a bit more accurate about which credentials we
store and the `receivedThisSessionFlag`.

For an operational profile reuse scenario, we can correctly set the flag
when a user is in migration - they will have received the operational
proile VC in the same session.

For a 6mfc journey we can just store the required credentials.
Previously we were storing all of them and would have had to remove the
fraud VC. Now it's just one operation.

For other reuse journeys we can store the credentials related to the
profile met - in practice this will just be gpg45 credentials.

Wynndow committed:
PYIC-5685: Simplify SessionCredentialService API

Rather than having two methods that both store credentials, we can just
have one. If we just want to store a single credential, we can wrap it
in a list.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -431,29 +438,43 @@ private JourneyResponse buildReuseResponse(
             // migration VC
-            return vcReceivedThisSession == null || vcReceivedThisSession.isEmpty()
+            boolean isOpProfileReuse =
+                    vcReceivedThisSession == null || vcReceivedThisSession.isEmpty();
```
We might need to rethink this a bit once we do the second phase - I was hoping we could remove `vcReceivedThisSession` entirely...

(No need to worry about backwards compat since operational profiles aren't live)

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -431,29 +438,43 @@ private JourneyResponse buildReuseResponse(
+        sessionCredentialsService.persistCredentials(
+                VcHelper.filterVCBasedOnProfileType(vcs, attainedVot.getProfileType()),
+                auditEventUser.getSessionId(),
+                false);
```
I wonder slightly if we could simplify by persisting at the top of the method and always filtering by a) profile type and b) removing any expired fraud VCs?

This is also fine, it just feels awkward to have the repetition.

Wynndow replied:
Yeah that's fair - i'll do it in a follow up PR though as you're away today and the PR is approved.

Joe-Edwards-GDS reviewed (approved):
We might still be able to tidy up the storing a little bit, but happy to go as-is if it doesn't work.
